python/cudaq/operator/cudm_state.py

```diff
@@ -0,0 +1,175 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+from cuquantum.densitymat import DenseMixedState, DensePureState, WorkStream
+import numpy, cupy, atexit
+from typing import Sequence
+from cupy.cuda.memory import MemoryPointer, UnownedMemory
+from ..mlir._mlir_libs._quakeDialects import cudaq_runtime
+
+
+# Wrap state data (on device memory) as a `cupy` array.
+# Note: the `cupy` array only holds a reference to the GPU memory buffer, no copy.
+def to_cupy_array(state):
+    tensor = state.getTensor()
+    pDevice = tensor.data()
+    dtype = cupy.complex128
+    sizeByte = tensor.get_num_elements() * tensor.get_element_size()
+    # Use `UnownedMemory` to wrap the device pointer
+    mem = UnownedMemory(pDevice, sizeByte, owner=state)
+    memptr = MemoryPointer(mem, 0)
+    cupy_array = cupy.ndarray(tensor.get_num_elements(),
+                              dtype=dtype,
+                              memptr=memptr)
+    return cupy_array
+
+
+# Helper to convert a state vector to a density matrix
+def ket2dm(ket: cupy.ndarray) -> cupy.ndarray:
+    return cupy.outer(ket.reshape((ket.size, 1)),
+                      cupy.conjugate(ket.reshape((ket.size, 1))))
+
+
+# Helper to create a 'coherent' state as a state vector.
+def coherent_state(N: int, alpha: float):
+    sqrtn = cupy.sqrt(cupy.arange(0, N, dtype=cupy.complex128))
+    sqrtn[0] = 1
+    data = alpha / sqrtn
+    data[0] = cupy.exp(-cupy.abs(alpha)**2 / 2.0)
+    cupy.cumprod(data, out=sqrtn)
+    return sqrtn
+
+
+# Helper to create a coherent state as a density matrix.
+def coherent_dm(N: int, alpha: float):
+    return ket2dm(coherent_state(N, alpha))
+
+
+# Evaluate the Wigner functions input state
+def wigner_function(state: cudaq_runtime.State | cupy.ndarray,
```

Comment:
This (and other functions) is literally a copy of QuTiP source code without acknowledging it under the BSD-3 license

Reply:
Thanks for the spotting this. 
This function was meant for internal testing of an example to verify against QuTiP, which should have been cleaned up. I've [removed](https://github.com/NVIDIA/cuda-quantum/pull/2391/commits/f80e6153a00c8a32a62d5726ad6981ec00b88f45) it as such.
Please let us know if you think other functions missing the acknowledgement.